src/main/java/gregtech/common/datafix/WorldDataHooks.java

```diff
@@ -0,0 +1,198 @@
+
+    private static final Version V1_10_5 = new Version(1, 10, 5); // granite was added in 1.10.5
+    private static final Version V1_14_0 = new Version(1, 14, 0); // meta block id alloc was changed in 1.14.0
+
+    private static int gtFallbackVersion = -1;
+    private static final TIntIntMap oldIdMapCompressed = new TIntIntHashMap(32, 0.8F, -1, -1);
+    private static final TIntIntMap oldIdMapCompressedInv = new TIntIntHashMap(32, 0.8F, -1, -1);
+    private static final TIntIntMap newIdMapCompressed = new TIntIntHashMap(64, 1F, -1, -1);
+    private static final TIntIntMap oldIdMapSurfaceRock = new TIntIntHashMap(4, 0.8F, -1, -1);
+    private static final TIntIntMap newIdMapSurfaceRock = new TIntIntHashMap(64, 1F, -1, -1);
+
+    public static void onWorldLoad(SaveHandler saveHandler, NBTTagCompound levelTag) {
+        gtFallbackVersion = -2;
+        oldIdMapCompressed.clear();
+        oldIdMapCompressedInv.clear();
+        newIdMapCompressed.clear();
+        oldIdMapSurfaceRock.clear();
+        newIdMapSurfaceRock.clear();
+
+        // read from gt world-saved data
+        File mapStorageFile = saveHandler.getMapFileFromName(MAP_STORAGE_NAME);
+        NBTTagCompound wsdDataTag = null;
+        if (mapStorageFile.exists()) {
+            try (FileInputStream mapStorageIn = new FileInputStream(mapStorageFile)) {
+                wsdDataTag = CompressedStreamTools.readCompressed(mapStorageIn).getCompoundTag("data");
+                if (wsdDataTag.hasKey(KEY_FALLBACK_DATA_VERSION, Constants.NBT.TAG_INT)) {
+                    gtFallbackVersion = wsdDataTag.getInteger(KEY_FALLBACK_DATA_VERSION);
+                    GTLog.logger.info("Using fallback data version {} from WSD", gtFallbackVersion);
+                }
+            } catch (IOException e) {
+                throw new IllegalStateException("Failed to read GregTech world-saved data!", e);
+            }
+        }
+
+        // that failing, infer from previously-saved map version
+        if (gtFallbackVersion < -1 && levelTag.hasKey("FML", Constants.NBT.TAG_COMPOUND)) {
+            NBTTagCompound fmlTag = levelTag.getCompoundTag("FML");
+            if (fmlTag.hasKey("ModList", Constants.NBT.TAG_LIST)) {
+                NBTTagList modListTag = fmlTag.getTagList("ModList", Constants.NBT.TAG_COMPOUND);
+                for (int i = 0; i < modListTag.tagCount(); i++) {
+                    NBTTagCompound modEntryTag = modListTag.getCompoundTagAt(i);
+                    if (modEntryTag.getString("ModId").equals(GTValues.MODID)) {
+                        Version version = Version.parse(modEntryTag.getString("ModVersion"));
+                        if (version.compareTo(V1_10_5) < 0) {
+                            gtFallbackVersion = -1;
+                        } else if (version.compareTo(V1_14_0) < 0) {
+                            gtFallbackVersion = 0;
+                        } else {
+                            gtFallbackVersion = 1;
+                        }
```

Comment:
this code is only necessary because we did not have data fixers in place when we published the version with the granite ID shift; this is just a hack that retroactively assigns a data version for those versions. as such, i don't think it makes much sense to move it into a dictionary or other data structure, since we know for certain that it'll never grow larger than these two comparisons

Reply:
Do I understand correctly that we won't need to add another if add new fixer. Because we will already have version with fixers, right?